onnxruntime/contrib_ops/cuda/transformers/generation_cuda_impl.cu

```diff
@@ -0,0 +1,775 @@
+                              int *d_values_out,
+                              int num_items,
+                              int num_segments,
+                              int *d_offsets,
+                              cudaStream_t stream,
+                              bool is_descending);
+
+template void LaunchSortPairs(void *d_temp_storage,
+                              size_t temp_storage_bytes,
+                              const half *d_keys_in,
+                              half *d_keys_out,
+                              const int *d_values_in,
+                              int *d_values_out,
+                              int num_items,
+                              int num_segments,
+                              int *d_offsets,
+                              cudaStream_t stream,
+                              bool is_descending);
+
+template <typename T>
+__global__ void FilterLogitsKernelCustom(float* d_sorted_logits_in,
+                                         const int* d_sorted_indices,
+                                         T* d_logits_in_out,
+                                         float top_p_threshold,
+                                         float filter_value,
+                                         int batch_size,
+                                         int vocab_size) {
+  int index = blockIdx.x * blockDim.x + threadIdx.x;
+
+  if (index >= batch_size * vocab_size) {
+    return;
+  }
+
+  int vocab_idx = index % vocab_size;
+  int batch_id = index / vocab_size;
+  int start_index = batch_id * vocab_size;
+
+  int count = vocab_idx;
+  float sum = 0.0f;
+  while (count >= 0) {
+    sum += d_sorted_logits_in[start_index];
+    ++start_index;
+    --count;
+  }
+
+  if (sum > top_p_threshold) {
+    // Shift the indices to the right by one according to the custom implementation.
+    int shifted_index = index + 1;
+    if (shifted_index % vocab_size != 0) {
+      int original_index = batch_id * vocab_size + d_sorted_indices[shifted_index];
+      d_logits_in_out[original_index] = (T)filter_value;
+    }
+  }
+}
+
+template <typename T>
+__global__ void FilterLogitsKernel(float* d_sorted_logits_in,
+                                   const int* d_sorted_indices,
+                                   T* d_logits_in_out,
+                                   float top_p_threshold,
+                                   float filter_value,
+                                   int min_tokens_to_keep,
+                                   int batch_size,
+                                   int vocab_size) {
+  int index = blockIdx.x * blockDim.x + threadIdx.x;
+
+  if (index >= batch_size * vocab_size) {
+    return;
+  }
+
+  int vocab_idx = index % vocab_size;
+  int batch_id = index / vocab_size;
+  int start_index = batch_id * vocab_size;
+
+  int count = vocab_idx;
+  float sum = 0.0f;
+  while (count >= 0) {
+    sum += d_sorted_logits_in[start_index];
+    ++start_index;
+    --count;
+  }
```

Comment:
it looks like quite slow. We need following PRs to optimize it. One thing i can think of now is to use cub's scan.

Reply:
will add a TODO